examples/pytorch/seal_ogbl/main.py

```diff
@@ -0,0 +1,426 @@
+import argparse
+import time
+import os, sys
+from tqdm import tqdm
+from sklearn.metrics import roc_auc_score
+import torch
+import dgl
+from torch.nn import BCEWithLogitsLoss
+from torch.utils.data import Dataset, DataLoader
+from ogb.linkproppred import DglLinkPropPredDataset, Evaluator
+from utils import *
+from models import *
+
+
+class SEALOGBLDataset(Dataset):
+    def __init__(self, root, graph, split_edge, num_hops, percent=100, split='train',
+                 ratio_per_hop=1.0, directed=False, dynamic=True) -> None:
+        super().__init__()
+        self.root = root
+        self.graph = graph
+        self.split = split
+        self.split_edge = split_edge
+        self.num_hops = num_hops
+        self.percent = percent
+        self.ratio_per_hop = ratio_per_hop
+        self.directed = directed
+        self.dynamic = dynamic
+
+        if 'weights' in self.graph.edata:
+            self.edge_weights = self.graph.edata['weights']
+        else:
+            self.edge_weights = None
+        if 'feat' in self.graph.ndata:
+            self.node_features = self.graph.ndata['feat']
+        else:
+            self.node_features = None
+
+        if not self.dynamic:
+            self.examples = self.load_cached()
+            return
+
+        pos_edge, neg_edge = get_pos_neg_edges(split, self.split_edge, self.graph, self.percent)
+        self.links = torch.cat([pos_edge, neg_edge], 0).tolist() # [Np + Nn, 2]
+        self.labels = [1] * len(pos_edge) + [0] * len(neg_edge)
+
+    def __len__(self):
+        return len(self.links)
+
+    def __getitem__(self, idx):
+        if not self.dynamic:
+            return self.examples[idx]
+
+        src, dst = self.links[idx]
+        y = self.labels[idx]
+        subg = k_hop_subgraph(src, dst, self.num_hops, self.graph,
+            self.ratio_per_hop, self.directed)
+
+        # remove the link between src and dst
+        if subg.has_edges_between(0, 1):
+            subg.remove_edges(subg.edge_ids(0, 1))
+        if subg.has_edges_between(1, 0):
+            subg.remove_edges(subg.edge_ids(1, 0))
+
+        NIDs, EIDs = subg.ndata[dgl.NID], subg.edata[dgl.EID]
+
+        z = drnl_node_labeling(subg.adj(scipy_fmt='csr'), 0, 1)
+        edge_weights = self.edge_weights[EIDs] if self.edge_weights is not None else None
+        x = self.node_features[NIDs] if self.node_features is not None else None
+
+        subg_aug = subg.add_self_loop()
```

Comment:
With PR #4261, we are able to save L71-73.

Reply:
not available in the latest release yet. should we amend it now?

Reply:
Ok. Let's keep it till the next release.